runtime/src/accounts.rs

```diff
@@ -550,49 +550,91 @@ impl Accounts {
+    pub fn commit_credits_and_rents_unsafe(
+        &self,
+        rent_collector: &RentCollector,
+        ancestors: &HashMap<Fork, usize>,
+        fork: Fork,
+    ) -> (u64, u64) {
         // Clear the credit only hashmap so that no further transactions can modify it
         let mut w_credit_only_account_locks = self.credit_only_account_locks.write().unwrap();
         let w_credit_only_account_locks =
             Self::get_write_access_credit_only(&mut w_credit_only_account_locks)
                 .expect("Credit only locks didn't exist in commit_credits");
-        self.store_credit_only_credits(w_credit_only_account_locks.drain(), ancestors, fork);
+        self.store_credit_only_credits_and_rents(
+            w_credit_only_account_locks.drain(),
+            rent_collector,
+            ancestors,
+            fork,
+        )
     }
 
-    fn store_credit_only_credits<I>(
+    fn store_credit_only_credits_and_rents<I>(
         &self,
         credit_only_account_locks: I,
+        rent_collector: &RentCollector,
         ancestors: &HashMap<Fork, usize>,
         fork: Fork,
-    ) where
+    ) -> (u64, u64)
+    where
         I: IntoIterator<Item = (Pubkey, CreditOnlyLock)>,
     {
-        for (pubkey, lock) in credit_only_account_locks {
-            let lock_count = *lock.lock_count.lock().unwrap();
-            if lock_count != 0 {
-                warn!(
-                    "dropping credit-only lock on {}, still has {} locks",
-                    pubkey, lock_count
-                );
-            }
-            let credit = lock.credits.load(Ordering::Relaxed);
-            if credit > 0 {
-                let mut account = self
-                    .load_slow(ancestors, &pubkey)
-                    .map(|(account, _)| account)
-                    .unwrap_or_default();
+        let mut accounts: Vec<(Pubkey, Account)> = vec![];
+        let mut total_rent_collected = 0;
+        let mut total_credit_credited = 0;
+
+        {
+            let accounts_index = self.accounts_db.accounts_index.read().unwrap();
+            let storage = self.accounts_db.storage.read().unwrap();
+            for (pubkey, lock) in credit_only_account_locks {
+                let lock_count = *lock.lock_count.lock().unwrap();
+                if lock_count != 0 {
+                    warn!(
+                        "dropping credit-only lock on {}, still has {} locks",
+                        pubkey, lock_count
+                    );
+                }
+                let credit = lock.credits.load(Ordering::Relaxed);
+
+                let (mut account, _) =
+                    AccountsDB::load(&storage, ancestors, &accounts_index, &pubkey)
+                        .unwrap_or_default();
                 account.lamports += credit;
-                self.store_slow(fork, &pubkey, &account);
+                total_credit_credited += credit;
+
+                if let Some((account, rent)) = rent_collector.update(account) {
```

Comment:
there's a bug here...  rent_collector should take an account ref and return a tuple of (Option<&Account>,u64)

the reason is that we may collect *some* rent and exhaust the account's lamports.  Also, pass-by-value is copying data, which is *bad*.

https://github.com/solana-labs/solana/blob/6123d2f9e8ea73356a9f5d641b70bfdc5f60536b/runtime/src/rent_collector.rs#L57

Reply:
Ok. Added a feature in `rent_collector` which allows us to collect some rent. 
Right now it is switched off with boolean. 

And also of course `rent_collector.update` takes a mutable ref and returns a ref.

Reply:
ok, it could also return a Result<u64, u64> ;)  whatever makes load_accounts() look best

that's in a separate PR?

Reply:
@rob-solana No, it is in this PR. Just another condition in `rent_collector.update`.
Didn't think functionality is that big to be included in separate PR.

Reply:
@rob-solana on hindsight, it should go in separate PR. (the partial rent deduction stuff, I mean).
Will undo changes from here.

Reply:
cool, yeah. I think that'd be better, thanks.

Reply:
done